### PR TITLE
docs: blog post — Context Engineering in Plugins & Ecosystem Squad

### PR DESCRIPTION
**Pull request by fdelbrayelle**
## Summary

- New blog post at `src/contents/blogs/context-engineering-plugins-squad/index.md`
- URL (no date): `/blogs/context-engineering-plugins-squad`
- Publication date: June 25, 2026
- Author: François Delbrayelle

## What the post covers

Introduces Context Engineering as adopted by the Plugins & Ecosystem Squad. Covers:
- GitHub issues as machine-readable specs (entrypoint for humans and agents)
- AI maturity framework (L1–L5b, squad operates at L4a)
- The 5-phase workflow: planning → approval gate → batch implementation → QA loop → ship
- Productivity metrics: ~8× speedup, ~30 min vs 4h, ~$1.65/issue
- How knowledge is structured as versioned Skills in a central hub
- The human role after Context Engineering (spec authoring, plan approval, PR review, release)
- Links to the DevLille 2026 talk repo: https://github.com/fdelbrayelle/workflow-superpowers

## Notes

- `main.png` is a placeholder — replace before publish
- No schema/FAQ block in this post (opinion/methodology piece, not a product feature)

## Test plan

- [ ] Check frontmatter parses correctly (title, date, category, author)
- [ ] Verify URL `/blogs/context-engineering-plugins-squad` resolves (no date segment)
- [ ] Replace placeholder `main.png` with actual cover image before publication

### CODE REVIEW

File: src/contents/blogs/context-engineering-plugins-squad/index.md [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+  image: fdelbrayelle
+  role: Lead Software Engineer
+image: ./main.png
+---
+
+At Kestra, building a new plugin feature involves the same repeatable steps every time: read the issue, design an approach, write the code, run the tests, open a pull request, get it reviewed, run QA, and ship. A senior engineer can do this in around four hours for a basic or medium-complexity task — and several days for a comprehensive, multi-task plugin — not because the work is hard, but because it is thorough.
+
+We asked a different question: what if we kept the thoroughness and removed the repetition?
+
+It started with a single experiment: a QA Skill to test whether structured instructions could reliably drive browser automation against a live Kestra instance. It worked. That early success seeded the rest — the planning Skill, then the developer and reviewer agents, and eventually the full orchestrated cycle described in this post.
```
Comment (wrussell1999) [inline src/contents/blogs/context-engineering-plugins-squad/index.md]: Feel like this should be capitalised as it’s a name?

```suggestion
It started with a single experiment: a QA Skill to test whether structured instructions could reliably drive browser automation against a live Kestra instance. It worked. That early success seeded the rest — the Planning Skill, then the developer and reviewer agents, and eventually the full orchestrated cycle described in this post.
```

File: src/contents/blogs/context-engineering-plugins-squad/index.md [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+
+I presented this approach at [DevLille 2026](https://github.com/fdelbrayelle/workflow-superpowers), with live terminal recordings of a full cycle on a [real Kestra issue](https://github.com/kestra-io/plugin-datagen/issues/53).
+
+## Why Prompt Engineering Is Not Enough
+
+Prompt engineering is about crafting the right question. Context Engineering is about building the right environment.
+
+A well-prompted agent can write a Kestra task that compiles. A context-engineered agent can write one that follows Kestra plugin conventions, passes the test suite, handles edge cases correctly, includes YAML usage examples, and creates a pull request with the right reviewer team and a valid issue link in the body — on the first attempt.
+
+The difference is not cleverness. It is structure: explicit domain knowledge encoded as **Skills** (focused markdown instruction sets under 500 lines each), deterministic workflow steps with clear success conditions, feedback loops that route QA failures back to the developer agent for correction, and **human approval gates** at the decisions that create the most value.
```
Comment (wrussell1999) [inline src/contents/blogs/context-engineering-plugins-squad/index.md]: “It’s not this, it’s that” sounds quite AI written

File: src/contents/blogs/context-engineering-plugins-squad/index.md [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+This dual-audience design is intentional. A human engineer can read the issue and understand what to build. An AI agent can parse the same file and extract acceptance criteria, affected plugin classes, edge cases to handle, and expected outputs. No ambiguity. No "we'll figure it out during implementation."
+
+The issue is not a ticket. It is the contract — and it drives the entire software development lifecycle (SDLC) that follows.
+
+## The SDLC: Who Does What, and When
+
+:::alert{type="info"}
+**Skills and agents are both plain markdown files** — but they behave differently. They run on [Claude Code](https://claude.ai/code) and [OpenCode](https://opencode.ai), with a build step that generates both formats from a single markdown source.
+
+A **Skill** is a procedural instruction set: a numbered sequence of steps, decision points, and success conditions that runs in the main context window. It is invoked by a human (e.g. `/kestra-plugin-planning`) and may orchestrate other steps or spawn agents. Think of it as a runbook the AI follows on your behalf.
```
Comment (wrussell1999) [inline src/contents/blogs/context-engineering-plugins-squad/index.md]: I wonder if you can simplify the Skills explanation earlier on as you go into more detail here. Skills comes up a few times with a bit of a description but this is repetitive IMO

File: src/contents/blogs/context-engineering-plugins-squad/index.md [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+
+An **agent** is a role definition: a system prompt that describes a specific persona, its responsibilities, and its constraints. When a Skill spawns an agent, that agent runs in its own isolated context window — no shared history with the orchestrator — and returns a structured result when done. The developer and the reviewer are agents; the planning and implementation steps are Skills.
```
Comment (wrussell1999) [inline src/contents/blogs/context-engineering-plugins-squad/index.md]: Agent is lower case but Skill is upper case?

File: src/contents/blogs/context-engineering-plugins-squad/index.md [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+
+The squad member runs `/kestra-plugin-managing-issues` with a description or an existing issue URL. The Skill first asks about the origin of the work — customer signal, Dev Marketing initiative, internal bet, or innovation — then asks whether this is a new feature, a bug fix, or a reformat of an existing issue. From those answers, it generates a fully structured GitHub issue body following Kestra plugin conventions — acceptance criteria, YAML examples, affected tasks, and a clear statement of why the work exists — and posts it to the correct repository.
+
+The issue becomes the contract for everything that follows. Weak spec in, weak output out.
+
+### Step 2 — Generate the Plan `/kestra-plugin-planning`
+
+**Actor: any squad member (human) → agent**
+
+The squad member runs `/kestra-plugin-planning` with the issue URL. A planning agent reads the issue, queries the [Kestra MCP server](/blogs/kestra-mcp-docs) for relevant task schemas, blueprints, and documentation, and generates a structured implementation plan covering four sections:
```
Comment (wrussell1999) [inline src/contents/blogs/context-engineering-plugins-squad/index.md]: Make the blog link relative

File: src/contents/blogs/context-engineering-plugins-squad/main.png
Comment (wrussell1999) [inline src/contents/blogs/context-engineering-plugins-squad/main.png:1]: Incorrect image (something is in the works I believe)

Comment (fdelbrayelle) [reply src/contents/blogs/context-engineering-plugins-squad/main.png:1]: Yes @Nico-Kestra is aware of it :)

File: src/contents/blogs/context-engineering-plugins-squad/index.md (hunk 4 above)
Comment (fdelbrayelle) [reply src/contents/blogs/context-engineering-plugins-squad/index.md]: Yeah I took example on Claude Code page about [Skills' best practices](https://platform.claude.com/docs/en/agents-and-tools/agent-skills/best-practices) where it seems to appear as a "mark" or emphasized concept.

File: src/contents/blogs/context-engineering-plugins-squad/index.md (hunk 5 above)
Comment (fdelbrayelle) [reply src/contents/blogs/context-engineering-plugins-squad/index.md]: It's already relative, isn't it?

File: src/contents/blogs/context-engineering-plugins-squad/index.md (hunk 3 above)
Comment (aj-emerich) [reply src/contents/blogs/context-engineering-plugins-squad/index.md]: Our Alert blocks are really ugly, and this one is very huge on the post. I would try to minimize the amount of text you include here, if possible.

Comment (aj-emerich) [reply src/contents/blogs/context-engineering-plugins-squad/index.md]: Could have to do with the fact that line breaks don't render in.

File: src/contents/blogs/context-engineering-plugins-squad/index.md [hunk 6]
```diff
@@ -0,0 +1,335 @@
...
+
+- **Design** — architectural approach and tradeoffs
+- **Tasks** — implementation steps as a checklist
+- **Edge Cases** — boundary conditions the implementation must handle
+- **Docs Impact** — whether user-facing documentation needs updating
+
+A plan is not always the output. Before generating one, the Skill triages the issue: if it looks like a usage problem — a misconfiguration, a missing property, a flow that can be fixed without touching plugin code — the Skill queries the [Kestra MCP server](/blogs/kestra-mcp-docs) for the relevant documentation and blueprints, and posts a fix attempt directly as a comment on the issue instead. No plan, no implementation cycle, no `/plan-approved` required. The reporter gets unblocked immediately.
```
Comment (wrussell1999) [inline src/contents/blogs/context-engineering-plugins-squad/index.md]: Make the links go to the markdown file. This should be better for website performance.

```suggestion
The squad member runs `/kestra-plugin-planning` with the issue URL. A planning agent reads the issue, queries the [Kestra MCP server](../kestra-mcp-docs/index.md) for relevant task schemas, blueprints, and documentation, and generates a structured implementation plan covering four sections:

- **Design** — architectural approach and tradeoffs
- **Tasks** — implementation steps as a checklist
- **Edge Cases** — boundary conditions the implementation must handle
- **Docs Impact** — whether user-facing documentation needs updating

A plan is not always the output. Before generating one, the Skill triages the issue: if it looks like a usage problem — a misconfiguration, a missing property, a flow that can be fixed without touching plugin code — the Skill queries the [Kestra MCP server](../kestra-mcp-docs/index.md) for the relevant documentation and blueprints, and posts a fix attempt directly as a comment on the issue instead. No plan, no implementation cycle, no `/plan-approved` required. The reporter gets unblocked immediately.
```

File: src/contents/blogs/context-engineering-plugins-squad/index.md [hunk 7]
```diff
@@ -0,0 +1,335 @@
+---
+title: "Context Engineering in Practice: Automating the Plugin SDLC at Kestra"
+description: "How Kestra's Plugins & Ecosystem Squad introduced Context Engineering to automate the full software development lifecycle (SDLC) for plugins — from GitHub issue to merged PR — cutting delivery time from 4 hours — and even way more — to 30 minutes using structured AI agents and human-in-the-loop approval gates."
```
Comment (wrussell1999) [inline src/contents/blogs/context-engineering-plugins-squad/index.md]: **Em dashes in frontmatter description** — Guidelines prohibit em dashes; use commas or parentheses. Also "and even way more" reads as filler; just cut it.

Suggested: `"...cutting delivery time from 4 hours (sometimes several days) to 30 minutes using structured AI agents and human-in-the-loop approval gates."`

File: src/contents/blogs/context-engineering-plugins-squad/index.md [hunk 8]
```diff
@@ -0,0 +1,335 @@
...
+author:
+  name: François Delbrayelle
+  linkedin: https://www.linkedin.com/in/fdelbrayelle/
+  twitter: "@fdelbrayelle"
+  image: fdelbrayelle
+  role: Lead Software Engineer
+image: ./main.png
+---
+
+At Kestra, building a new plugin feature involves the same repeatable steps every time: read the issue, design an approach, write the code, run the tests, open a pull request, get it reviewed, run QA, and ship. A senior engineer can do this in around four hours for a basic or medium-complexity task — and several days for a comprehensive, multi-task plugin — not because the work is hard, but because it is thorough.
```
Comment (wrussell1999) [inline src/contents/blogs/context-engineering-plugins-squad/index.md]: **Em dashes (4 on one line)** — This line has four em dashes, which is the single biggest pattern violation in the piece. Guidelines say to use commas or parentheses instead.

Suggested: `"...in around four hours for a basic or medium-complexity task (and several days for a comprehensive, multi-task plugin), not because the work is hard, but because it is thorough."`

Note: the "not because X, but because Y" construction at the end is also flagged by the guidelines as negative parallelism. Consider: `"...not because the work is hard. It is thorough."`

File: src/contents/blogs/context-engineering-plugins-squad/index.md [hunk 9]
```diff
@@ -0,0 +1,335 @@
...
+
+It started with a single experiment: a QA Skill to test whether structured instructions could reliably drive browser automation against a live Kestra instance. It worked. That early success seeded the rest — the Planning Skill, then the developer and reviewer agents, and eventually the full orchestrated cycle described in this post.
+
+This post describes how the Plugins & Ecosystem Squad introduced **Context Engineering** into its development workflow, and what happened when we let AI agents handle the mechanical parts while humans stayed in control of the decisions that actually matter.
+
+I presented this approach at [DevLille 2026](https://github.com/fdelbrayelle/workflow-superpowers), with live terminal recordings of a full cycle on a [real Kestra issue](https://github.com/kestra-io/plugin-datagen/issues/53).
+
+## Why Prompt Engineering Is Not Enough
+
+Prompt engineering is about crafting the right question. Context Engineering is about building the right environment.
```
Comment (wrussell1999) [inline src/contents/blogs/context-engineering-plugins-squad/index.md]: **Negative parallelism** — "Prompt engineering is about crafting the right question. Context Engineering is about building the right environment." This is the classic "not X, it's Y" pattern in disguise (parallel negation without the explicit negation word). The contrast is clear without the rhetorical setup. Consider collapsing into a single direct statement: "Context Engineering is about building the right environment, not just crafting the right question."

Or drop the framing entirely and open directly with what Context Engineering *is*.

File: src/contents/blogs/context-engineering-plugins-squad/index.md [hunk 10]
```diff
@@ -0,0 +1,335 @@
...
+|-------|------|-------------|
+| L1 | Manual | Prompts typed into a chat UI. Copy-paste driven. |
+| L2 | Augmented | Copilot suggestions inline while a human codes. |
+| L3 | Scripted | Deterministic, human-designed automation — fixed steps, no agent autonomy. |
+| **L4a** | **Supervised Agentic** | **Dynamic multi-agent workflows with explicit human approval gates.** |
+| L4b | Autonomous Agentic | Agents run end-to-end without human checkpoints. |
+| L5a | Self-Optimizing | Agents emit telemetry and propose improvements to their own Skills. |
+| L5b | Self-Authoring | Agents create new agents and Skills from scratch. |
+
+The Plugins & Ecosystem Squad operates at **L4a**: agents handle execution, humans own decisions. Every approval gate is an explicit checkpoint — not an accidental pause.
```
Comment (wrussell1999) [inline src/contents/blogs/context-engineering-plugins-squad/index.md]: **Em dash** — "Every approval gate is an explicit checkpoint — not an accidental pause." Replace the em dash with a comma: "Every approval gate is an explicit checkpoint, not an accidental pause."

File: src/contents/blogs/context-engineering-plugins-squad/index.md [hunk 11]
```diff
@@ -0,0 +1,335 @@
...
+
+Most teams skip directly from L2 to L3 and stall there, because scripted automation breaks when requirements change. The jump to L4a — where agents adapt dynamically within a structured context — is where real productivity gains begin.
+
+## GitHub Issues as Machine-Readable Specs
+
+The workflow starts where all feature work starts: a GitHub issue. But before any issue is written, a more fundamental question has to be answered — **why does this work exist?**
+
+Every plugin feature on the backlog has an origin. It is either:
+- A **customer signal** — a request surfaced by the Sales or Customer Success squads, with real names and real use cases attached
```
Comment (wrussell1999) [inline src/contents/blogs/context-engineering-plugins-squad/index.md]: **Bold-first bullets** — Guidelines flag this as a strong AI tell. Every bullet starts with a bolded label followed by a dash and explanation. Real editorial prose doesn't format lists this way.

Suggested: either drop the bold and integrate the label into the sentence naturally, or restructure as short declarative bullets without the label-dash pattern:
```
- A request surfaced by Sales or Customer Success, with real names and use cases attached
- A plugin that strengthens an integration story, a category, or a release theme
- A connector the squad believes is strategically valuable, before external demand has materialized
- An experiment or proof-of-concept, not yet externally requested
```

File: src/contents/blogs/context-engineering-plugins-squad/index.md [hunk 12]
```diff
@@ -0,0 +1,335 @@
...
+- A **Dev Marketing initiative** — a plugin that strengthens an integration story, a category, or a release theme
+- An **internal bet** — the squad believes a new connector or capability is strategically valuable, before external demand has materialized
+- **Innovation** — an experiment, a proof-of-concept, something nobody asked for yet
+
+This is product thinking, and it belongs at the very start of the software development lifecycle (SDLC). The answer shapes the spec: a customer-signal issue names the use case and the success criterion. A Dev Marketing issue frames the plugin in the context of the integration story it supports. An innovation issue admits its exploratory nature upfront and scopes the acceptance criteria accordingly.
+
+An issue that skips the "why" is an issue that may be built correctly but shipped to nobody.
+
+We write issues differently now. A good issue for this workflow is a complete business and technical specification in markdown. It includes:
```
Comment (wrussell1999) [inline src/contents/blogs/context-engineering-plugins-squad/index.md:83]: **Bold-first bullets again** — Same pattern as lines 68-71. Consider plain bullets or prose integration for at least one of these list sections.

File: src/contents/blogs/context-engineering-plugins-squad/index.md [hunk 13]
```diff
@@ -0,0 +1,335 @@
...
+- A description of the desired behavior
+- Acceptance criteria as a checklist
+- One or more YAML usage examples showing what a correct implementation looks like from the user's perspective
+
+This dual-audience design is intentional. A human engineer can read the issue and understand what to build. An AI agent can parse the same file and extract acceptance criteria, affected plugin classes, edge cases to handle, and expected outputs. No ambiguity. No "we'll figure it out during implementation."
+
+The issue is not a ticket. It is the contract — and it drives the entire SDLC that follows.
```
Comment (wrussell1999) [inline src/contents/blogs/context-engineering-plugins-squad/index.md]: **Negative parallelism** — "The issue is not a ticket. It is the contract." This is a textbook example of the prohibited pattern: sentence 1 negates, sentence 2 reframes. Guidelines flag this as the "single most commonly identified AI writing tell."

Suggested: "The issue is the contract, and it drives the entire SDLC that follows."

File: src/contents/blogs/context-engineering-plugins-squad/index.md [hunk 14]
```diff
@@ -0,0 +1,335 @@
...
+
+A plan is not always the output. Before generating one, the Skill triages the issue: if it looks like a usage problem — a misconfiguration, a missing property, a flow that can be fixed without touching plugin code — the Skill queries the [Kestra MCP server](/blogs/kestra-mcp-docs) for the relevant documentation and blueprints, and posts a fix attempt directly as a comment on the issue instead. No plan, no implementation cycle, no `/plan-approved` required. The reporter gets unblocked immediately.
+
+Only when the issue is clearly a plugin code change does the Skill generate and post the structured plan. The squad member reads it, asks questions if needed, and — when satisfied — posts exactly `/plan-approved` on the issue.
+
+### Step 3 — Approval Gate `/plan-approved`
+
+**Actor: any kestra-io org member (human)**
+
+This step does not require a Skill. The approver posts `/plan-approved` as a comment on the issue. That is all.
```
Comment (wrussell1999) [inline src/contents/blogs/context-engineering-plugins-squad/index.md]: **Short punchy fragment** — "That is all." is a manufactured-emphasis fragment. Guidelines flag this pattern specifically. Cut it; the preceding sentence already makes the point completely.

File: src/contents/blogs/context-engineering-plugins-squad/index.md [hunk 15]
```diff
@@ -0,0 +1,335 @@
...
+
+But the headline number understates the change. Manual processes routinely compress some steps under time pressure — planning gets light, code review skims, QA gets cut short. With agents, every issue gets thorough planning, detailed review across four tracks, and QA that covers happy paths, failure scenarios, and non-regression. The quality floor rises alongside the speed.
+
+### How We Keep the Cost Down
+
+Token consumption in agentic workflows grows faster than linearly: every turn adds to the context window that all subsequent turns must read. We use three techniques to counteract this.
+
+**Pass the plan to the developer.** The planning Skill maps the relevant classes and design decisions upfront and posts them as a structured comment on the issue. The developer agent reads that comment and targets the listed files directly — skipping broad codebase exploration. Eliminating even a few exploration turns has an outsized effect: each saved turn roughly halves its own cost because context window growth is roughly triangular.
```
Comment (wrussell1999) [inline src/contents/blogs/context-engineering-plugins-squad/index.md]: **Em dash** — "...before running tests), after the developer agent returns, and after QA (which accumulates significant browser and terminal history). Each compaction flushes..."

The surrounding text uses em dashes in several places in this paragraph (lines 254-260). Replace with commas or recast.

File: src/contents/blogs/context-engineering-plugins-squad/index.md [hunk 16]
```diff
@@ -0,0 +1,335 @@
...
+
+The architecture we built supports L5a: Skills & agents are version-controlled markdown files, agent output is structured and capturable, and the feedback loops already exist between QA, code review, and the developer agent. The missing piece is the telemetry layer and the self-improvement agent that reads it. That is the next thing we are building.
+
+---
+
+The deeper invitation here is not to copy this workflow — it is to look at your own squad and ask: what does your team know that an agent does not? What conventions, guardrails, and review instincts live only in people's heads? Write those down as Skills and agents. Test them on real issues. Iterate. The stack will keep changing; the knowledge your team encodes will not.
+
+What comes after context engineering is loop engineering: wrapping agents in verification, event-driven, and hill-climbing loops so the system not only executes but continuously improves itself — closing the gap between what the agent does today and what your team would have done instead.
```
Comment (wrussell1999) [inline src/contents/blogs/context-engineering-plugins-squad/index.md]: **Em dash + grandiose closing** — "closing the gap between what the agent does today and what your team would have done instead" ends with an em dash earlier in the sentence. Also "continuously improves itself" leans toward grandiose stakes inflation.

The final sentence is strong conceptually but could be tightened. Also check: the two-sentence conclusion ("The deeper invitation..." and "What comes after...") slightly fractures the landing. One crisp closing thought would land harder.

**Review by wrussell1999 — COMMENTED** [review]
Style review against my writing CLAUDE.md. Main issues: em dashes throughout (prohibited), negative parallelism pattern, bold-first bullets, and a few short punchy fragments used for manufactured emphasis.

File: src/contents/blogs/context-engineering-plugins-squad/main.png
Comment (fdelbrayelle) [reply src/contents/blogs/context-engineering-plugins-squad/main.png:1]: Acknowledged — this is a known placeholder. The cover image will be replaced before publication (noted in the PR description).

File: src/contents/blogs/context-engineering-plugins-squad/index.md (hunk 4 above)
Comment (fdelbrayelle) [reply src/contents/blogs/context-engineering-plugins-squad/index.md]: Deliberate convention: **Skill** is capitalized because it refers to a specific named product concept (the slash-command instruction files). **agent** stays lowercase because it's used both as a generic term ("planning agent", "sub-agents") and as a named role — capitalizing it everywhere would read oddly alongside phrases like "AI agents" or "agentic workflow". The callout box (lines 38–40) makes the distinction explicit.

File: src/contents/blogs/context-engineering-plugins-squad/index.md (hunk 3 above)
Comment (fdelbrayelle) [reply src/contents/blogs/context-engineering-plugins-squad/index.md]: Good call. The callout box (§ SDLC heading) was moved earlier in a prior commit to front-load the definition, but that created repetition with the 'How the Knowledge Is Structured' section. I'll trim the callout to a single-sentence pointer and leave the full structural explanation to the dedicated section.

File: src/contents/blogs/context-engineering-plugins-squad/index.md (hunk 1 above)
Comment (fdelbrayelle) [reply src/contents/blogs/context-engineering-plugins-squad/index.md]: Already capitalized in current text — this was fixed in a prior commit on this branch.

File: src/contents/blogs/context-engineering-plugins-squad/index.md (hunk 2 above)
Comment (fdelbrayelle) [reply src/contents/blogs/context-engineering-plugins-squad/index.md]: Fixed — collapsed to a single direct statement: 'Context Engineering is about building the right environment, not just crafting the right question.'

**Review by wrussell1999 — APPROVED** [review]

**Review by aj-emerich — APPROVED** [review]
